distributed/_ipython_utils.py

```diff
@@ -114,3 +178,72 @@ def _cleanup_connection_file():
             pass
     atexit.register(_cleanup_connection_file)
 
+
+def start_ipython(ip=None, ns=None, log=None):
+    """Start an IPython kernel in a thread
+
+    Parameters
+    ----------
+
+    ip: str
+        The IP address to listen on (likely the parent object's ip).
+    ns: dict
+        Any names that should be injected into the IPython namespace.
+    log: logger instance
+        Hook up IPython's logging to an existing logger instead of the default.
+    """
+    from IPython import get_ipython
+    if get_ipython() is not None:
+        raise RuntimeError("Cannot start IPython, it's already running.")
+
+    from zmq.eventloop.ioloop import ZMQIOLoop
+    from ipykernel.kernelapp import IPKernelApp
+    # save the global IOLoop instance
+    # since IPython relies on it, but we are going to put it in a thread.
+    save_inst = IOLoop.instance()
+    IOLoop.clear_instance()
+    zmq_loop = ZMQIOLoop()
+    zmq_loop.install()
+
+    # start IPython, disabling its signal handlers that won't work due to running in a thread:
+    app = IPKernelApp.instance(log=log)
+    # Don't connect to the history database
+    app.config.HistoryManager.hist_file = ':memory:'
+    # listen on all interfaces, so remote clients can connect:
+    if ip:
+        app.ip = ip
+    # disable some signal handling, logging
+    noop = lambda : None
+    app.init_signal = noop
+    app.log_connection_info = noop
+
+    # start IPython in a thread
+    # initialization happens in the thread to avoid threading problems
+    # with the sqlite history
+    evt = Event()
+    def _start():
+        app.initialize([])
+        app.kernel.pre_handler_hook = noop
+        app.kernel.post_handler_hook = noop
+        app.kernel.start()
+        app.kernel.loop = IOLoop.instance()
+        # save self in the IPython namespace as 'worker'
+        # inject things into the IPython namespace
+        if ns:
+            app.kernel.shell.user_ns.update(ns)
+        from tornado.ioloop import PeriodicCallback
+        pc = PeriodicCallback(lambda : print(os.getpid(), file=sys.__stdout__), 1000, io_loop=zmq_loop)
```

Comment:
super-professional print-statement debugging that I neglected to clean up. Gone now.